### DISCUSSION TITLE
colorOrder for individual LEDs

### DISCUSSION THREAD

**Discussion by Schnielz** [awawa-dev/HyperHDR#478] [Q&A]
Hello All,

I am using two different types of LEDs in the setup of my TV (because I had to buy more LEDs for a new bigger screen). Now the color order (RGB, RBG, etc.) of the LEDs differ. Color channel adjustments can be applied for each LED by using multiple profiles but the color order cannot be configured individually. It would be good to make it possible to configure it in the LED layout.

In the past it was possible to add a colorOrder property to each LED in the exported backup JSON file and re-import it. But somewhen between HyperHDR 17.0.0 and 18.0.0 the processig changed and only the last colorOrder configured in the string is applied to all the LEDs. The configured global "RGB byte order" is ignored in that case as well.

Has anybody an idea how to configure it? Or should I create a feature request?

**Comment by awawa-dev** [comment]
Hi
Send me your HyperHDR configuration exported backup. BTW for your setup, HyperSerialESP32 with multi-segment support should be a better option (could be compiled with each segment having different RGB order).

**Comment by Schnielz** [comment]
Hi
I am not sure what HyperSerialESP32 does and if it works with my WS2801 and Ezcap 269 setup.

Here you see my backup file. It works fine in HyperHDR 17.0.0 and used to work similar in HyperHDR 15.0.0. In 18.0.0 colorOrder rgb is used for all the LEDs. I played a bit around and it doesn't make a difference what colorOrder is defined in the device config or in the other LEDs. Only the last one in row 67 is applied to all the LEDs.

[HyperHDR-17.0.0.0-Backup-2023-01-05.json.txt](https://github.com/awawa-dev/HyperHDR/files/10355621/HyperHDR-17.0.0.0-Backup-2023-01-05.json.txt)

**Comment by awawa-dev** [comment]
Did you try to change the colors for that fragment of LED strip in the processing/calibration tab? e.g. change RGB of red value from (255,0,00) to green (0,255,0) ...swap all colors.

Probably it will work only for the new calibration, not for the 'classic' one with luminance/saturation etc

EDIT:
I check v17 code: that field colorOrder from the individual LED was completely ignored per LED, it overrides the whole strip	[ledString.colorOrder = stringToColorOrder(ledConfig["colorOrder"].toString(deviceOrderStr));](https://github.com/awawa-dev/HyperHDR/blob/5a8ff444c1d9dd73493b419eceb47e2e2164faee/sources/base/LedString.cpp#L53)

**Comment by Schnielz** [comment]
I was afraid to break my config by that but now I tried and the workaround does the trick.  Thanks for your help!

**Comment by Schnielz** [comment]
Hi, I just saw your edit.

Yes, the LED struct used to have its own colorOrder back then:
https://github.com/awawa-dev/HyperHDR/blob/9881cf82d0c33337ce98868d8173802edf6eb135/include/hyperhdrbase/LedString.h#L104

In the current version each processed ledConfig overrides the colorOrder of the whole string. Resulting in the fact that the last one is used only. I would like to see the colorOrder per Led back but the Led struct doesn't even have a colorOrder defined any longer. Might be a too big refatroing for this edge case. Cleaner at least could be to remove the line you mentioned above completely and assign deviceOrderStr only.
